server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1098,6 +1102,122 @@ private void asyncCleanupUnlinkedShardLevelBlobs(
     }
 
+    public static void remoteDirectoryCleanupAsync(
+        RemoteSegmentStoreDirectoryFactory remoteDirectoryFactory,
+        ThreadPool threadpool,
+        String remoteStoreRepoForIndex,
+        String indexUUID,
+        ShardId shardId
+    ) {
+        threadpool.executor(ThreadPool.Names.REMOTE_PURGE)
+            .execute(
+                new RemoteStoreShardCleanupTask(
+                    () -> RemoteSegmentStoreDirectory.remoteDirectoryCleanup(
+                        remoteDirectoryFactory,
+                        remoteStoreRepoForIndex,
+                        indexUUID,
+                        shardId
+                    ),
+                    threadpool.executor(ThreadPool.Names.REMOTE_PURGE),
+                    indexUUID,
+                    shardId
+                )
+            );
+    }
+
+    /**
+    A Runnable wrapper to make sure that for a given shard only one cleanup task runs at a time.
+     */
+    public static class RemoteStoreShardCleanupTask implements Runnable {
+        private final Runnable task;
+        private final ExecutorService executor;
+        private final String shardIdentifier;
+        final static Set<String> ongoingRemoteDirectoryCleanups = newConcurrentSet();
+        final static ConcurrentMap<String, BlockingQueue<Runnable>> shardCleanupPendingTasks = new ConcurrentHashMap<>();
+
+        public RemoteStoreShardCleanupTask(Runnable task, ExecutorService executor, String indexUUID, ShardId shardId) {
+            this.task = task;
+            this.shardIdentifier = indexShardIdentifier(indexUUID, shardId);
+            this.executor = executor;
+        }
+
+        private static String indexShardIdentifier(String indexUUID, ShardId shardId) {
+            return String.join("/", indexUUID, String.valueOf(shardId.id()));
+        }
+
+        @Override
+        public void run() {
+            if (ongoingRemoteDirectoryCleanups.add(shardIdentifier)) {
+                try {
+                    task.run();
+                    BlockingQueue<Runnable> pendingTasks = shardCleanupPendingTasks.get(shardIdentifier);
+                    try {
+                        if (pendingTasks != null) {
+                            for (Runnable pendingTask = pendingTasks.poll(0L, TimeUnit.MILLISECONDS); pendingTask != null;) {
+                                pendingTask.run();
+                            }
+                        }
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                } finally {
+                    ongoingRemoteDirectoryCleanups.remove(shardIdentifier);
+                }
+            } else {
+                shardCleanupPendingTasks.putIfAbsent(shardIdentifier, new LinkedBlockingQueue<>());
+                shardCleanupPendingTasks.get(shardIdentifier).add(task);
+            }
+        }
+    }
+
+    protected void releaseRemoteStoreLockAndCleanup(
```

Comment:
Given the remote store data deletion is async, we can go in a situation where the shard blob is deleted but the remote store data is not deleted. Can we use an approach where we delete the shard blobs only after the remote store clean up is done successfully. This, currently, puts us at the risk of zombie data and can lead to higher storage cost.

Reply:
that is not achievable at least right now.. since snapshot deletion is async. we need to implement GC component or some monitor which takes care of cleaning up zombie data.